packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,235 @@
+ * @param {string} searchString				 The string used as the base for the
+ *                                     search
+ * @param {Object} conversionConfig    The configuration object for the
+ *                                     interpolation being performed.
+ *
+ * @return {Array|null}	An array if there is a match or null if not.
+ */
+const getMatchFromString = (
+	interpolatedString,
+	searchString,
+	conversionConfig
+) => {
+	const tags = getTagsFromString( searchString );
+	// first try children reg ex.  If there is a match, then return.
+	const match = interpolatedString.match(
+		getRegEx( tags, searchString, 'children' )
+	);
+	if ( match !== null ) {
+		conversionConfig.hasChildren = true;
+		return match;
+	}
+	// no children, try matching on another pattern and return.
+	const regEx = getHasPropValue( conversionConfig ) ?
+		getRegEx( tags, searchString, 'propValue' ) :
+		getRegEx( tags, searchString );
+	return interpolatedString.match( regEx );
+};
+
+// index for keys
+// This is external to `recursiveCreateElement` and reset in
+// `createInterpolateElement` because of the recursion.
+let keyIndex = -1;
+
+/**
+ * Used to recursively create elements from the interpolation string using the
+ * conversion map.
+ *
+ * @param {string}  potentialElement  The interpolation string (or fragment)
+ *                                    being processed.
+ * @param {Array[]} conversionMap     The interpolation map used for converting
+ *                                    the string to a react element.
+ *
+ * @return {Element|string|Array}  A react element, string or array.
+ */
+const recursiveCreateElement = ( potentialElement, conversionMap ) => {
+	/**
+	 * If the conversion map is not a valid array or empty then just return the
+	 * element.
+	 */
+	if ( ! Array.isArray( conversionMap ) || ! conversionMap.length ) {
+		return potentialElement;
+	}
+	const [ mapItem ] = conversionMap.slice( 0, 1 );
+	const [ searchString, conversionConfig ] = mapItem;
+
+	/**
+	 * This short circuits the process if the conversion map has an invalid config.
+	 */
+	if ( ! searchString || ! conversionConfig ) {
+		return potentialElement;
+	}
+
+	const match = getMatchFromString(
+		potentialElement,
+		searchString,
+		conversionConfig
+	);
+
+	// if there is no match for this string, then that means it is not an element
+	// so just return as is to be used as a direct child.
+	if ( match === null ) {
+		return potentialElement;
+	}
+
+	// if the full match returned equals the potential element, then we know
+	// we can create the element and restart the conversion on any children if
+	// necessary.
+	if ( match[ 0 ] === potentialElement ) {
+		// remove this item from the conversion map because it's no longer needed.
+		conversionMap.shift();
+
+		if ( getHasPropValue( conversionConfig ) ) {
+			// if value is a react element, then need to wrap in Fragment with a key
+			// to prevent key warnings.
```

Comment:
* I think the keying isn't there so much to prevent the warnings, but rather to help with reconciliation itself. ;)

* Would it be better, to simplify the logic involved, to key these values regardless of being a valid element?

```js
if ( getHasPr… ) {
  return <Fragment key={ … }>{ conversionConfig.value }</Fragment>;
```

Reply:
That would wrap all primitive values (i.e. string or number) in a Fragment unnecessarily.  For example something like this:

```
__experimentalCreateInterpolateElement(
    'Some %1$s value and another %2$s number',
    { '%1$s': { value: 'concrete' }, '%2$s': { value: 30 } }
);
```

...would end up equivalent to:

```jsx
createElement(
  Fragment,
  {},
  [
    'Some ',
    createElement(
      Fragment,
      {},
      [ 'concrete' ]
    ),
    ' value and another ',
    createElement(
      Fragment,
      {},
      [ 30 ]
    ),
    ' number'
  ]
);
```

..instead of what we have currently which would be:

```jsx
createElement(
  Fragment,
  {},
  [
    'Some ',
    createElement(
      Fragment,
      {},
      [ 'concrete' ]
    ),
    ' value and another ',
    createElement(
      Fragment,
      {},
      [ 30 ]
    ),
    ' number'
  ]
);
```

So essentially, why wrap in `Fragment` if we don't need to?

Reply:
Indeed, it's not worth the messy output at all. I was trying to figure out ways to streamline `recursiveCreateElement`, but this isn't the way to go about it. :)